src/platform/test/mtest-main.c

```diff
@@ -0,0 +1,215 @@
+/* Copyright (c) 2022 Felix Jones
+*
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <mgba/core/cheats.h>
+#include <mgba/core/config.h>
+#include <mgba/core/core.h>
+#include <mgba/core/serialize.h>
+#include <mgba/internal/gba/gba.h>
+
+#include <mgba/feature/commandline.h>
+#include <mgba-util/vfs.h>
+
+#include <signal.h>
+
+#define MTEST_OPTIONS "S:R:O:"
+#define MTEST_USAGE \
+	"\nAdditional options:\n" \
+	"  -S SWI           Run until specified SWI call before exiting\n" \
+	"  -R REGISTER      Register to return as exit code\n" \
+
+ enum mRegisterExitCode {
+	 mRegister_NONE = -1,
+	 mRegister_GPRS = 0,
+	 mRegister_CPSR = 16
+ };
+
+struct MTestOpts {
+	int exitSwiImmediate;
+	enum mRegisterExitCode returnCodeRegister;
+};
+
+static int _mTestRunloop(struct mCore* core, int exitSwiImmediate);
+static void _mTestShutdown(int signal);
+static bool _parseMTestOpts(struct mSubParser* parser, int option, const char* arg);
+static enum mRegisterExitCode _parseNamedRegister(const char* regStr);
+
+static bool _dispatchExiting = false;
+
+#ifdef M_CORE_GBA
+void (*_armSwi16)(struct ARMCore* cpu, int immediate);
+void (*_armSwi32)(struct ARMCore* cpu, int immediate);
+
+static void _mTestSwi16(struct ARMCore* cpu, int immediate);
+static void _mTestSwi32(struct ARMCore* cpu, int immediate);
+static void _mTestUpdateExitCond(struct ARMCore* cpu, int immediate);
+
+static enum mRegisterExitCode _returnCodeRegister;
+static int _exitCode = 0;
+static int _prevSwiImmediate = -1;
+#endif
+
+int main(int argc, char * argv[]) {
+	signal(SIGINT, _mTestShutdown);
+
+	struct MTestOpts mTestOpts = { -1, mRegister_NONE };
+	struct mSubParser subparser = {
+		.usage = MTEST_USAGE,
+		.parse = _parseMTestOpts,
+		.extraOptions = MTEST_OPTIONS,
+		.opts = &mTestOpts
+	};
+
+	struct mArguments args;
+	bool parsed = parseArguments(&args, argc, argv, &subparser);
+	if (!args.fname) {
+		parsed = false;
+	}
+	if (!parsed || args.showHelp) {
+		usage(argv[0], MTEST_USAGE);
+		return !parsed;
+	}
+	if (args.showVersion) {
+		version(argv[0]);
+		return 0;
+	}
+	struct mCore* core = mCoreFind(args.fname);
+	if (!core) {
+		return 1;
+	}
+	core->init(core);
+	mCoreInitConfig(core, "mTest");
+	applyArguments(&args, NULL, &core->config);
+
+	mCoreConfigSetDefaultValue(&core->config, "idleOptimization", "remove");
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+		((struct GBA*) core->board)->cpu->irqh.swi16 = _mTestSwi16;
+		_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+		((struct GBA*) core->board)->cpu->irqh.swi32 = _mTestSwi32;
```

Comment:
Instead of hooking all SWIs and only using one, perhaps using SWI 03 (STOP) which already has a frontend-defined callback would be ideal here. See the `GBAStop` function in gba.c.

Reply:
I agree that GBAStop makes a perfect choice, however my own runtime uses SWI 0 to reset when exit is called. I've left in the option to override the SWI call, but if you absolutely disagree with this then I can adjust my runtime and remove this option here